src/protocol/sort/generate_sort_permutations.rs

```diff
@@ -0,0 +1,110 @@
+use crate::{
+    error::BoxError,
+    ff::Field,
+    protocol::{
+        context::ProtocolContext,
+        modulus_conversion::convert_shares::convert_all,
+        sort::bit_permutations::BitPermutations,
+        sort::SortStep::{BitPermutation, ModulusConversion},
+    },
+    secret_sharing::Replicated,
+};
+
+use super::{compose::Compose, secureapplyinv::SecureApplyInv};
+
+/// This is an implementation of `GenPerm` (Algorithm 6) described in:
+/// "An Efficient Secure Three-Party Sorting Protocol with an Honest Majority"
+/// by K. Chida, K. Hamada, D. Ikarashi, R. Kikuchi, N. Kiribuchi, and B. Pinkas
+/// <https://eprint.iacr.org/2019/695.pdf>.
+#[derive(Debug)]
+pub struct GenerateSortPermutations<'a> {
+    input: &'a [u64],
+}
+
+impl<'a> GenerateSortPermutations<'a> {
+    #[allow(dead_code)]
+    pub fn new(input: &'a [u64]) -> GenerateSortPermutations {
+        Self { input }
+    }
+
+    #[allow(dead_code)]
+    pub async fn execute<F: Field>(
+        &self,
+        ctx: ProtocolContext<'_, F>,
+    ) -> Result<Vec<Replicated<F>>, BoxError> {
+        let mut bits = convert_all(ctx.narrow(&ModulusConversion), self.input).await?;
+        let mut sigma = BitPermutations::new(&bits[0])
+            .execute(ctx.narrow(&BitPermutation))
+            .await?;
+        for (i, ith_bit) in bits.iter_mut().enumerate().skip(1) {
+            SecureApplyInv::execute(
```

Comment:
@akoshelev @martinthomson needed some help in what context needs to be called here for applyinv, bit permutation and compose. These three protocols will be called for each bit in the match key.